expan/core/experiment.py

```diff
@@ -20,396 +24,103 @@
-            kpis_to_analyse: list of KPIs to be analysed
-            distribution: name of the KPI distribution model, which assumes a
-                Stan model file with the same name exists
-            posterior_width: the stopping criterion, threshold of the posterior 
-                width
-            num_iters: number of iterations for the bayes sampling
+        if not method in workerTable:
+            raise NotImplementedError
 
-        Returns:
-            a Results object
-        """
-        for mname in kpis_to_analyse:
-            metric_df = self._get_metric_df(mname, reference_kpis, weighted_kpis)
-
-            def do_delta(x, y, x_weights, y_weights):
-                weighted_x = np.array(x, dtype=float) * x_weights
-                weighted_y = np.array(y, dtype=float) * y_weights
-                return early_stopping_to_dataframe(metric_df.columns[2],
-                                                   *es.bayes_precision(
-                                                       x=weighted_x,
-                                                       y=weighted_y,
-                                                       distribution=distribution,
-                                                       posterior_width=posterior_width,
-                                                       num_iters=num_iters))
-
-            df = metric_df.groupby('variant').apply(self._apply_reweighting_and_all_variants,
-                                                    metric_df=metric_df,
-                                                    weighted_kpis=weighted_kpis,
-                                                    reference_kpis=reference_kpis,
-                                                    mname=mname,
-                                                    func_apply_variants=do_delta).unstack(0)
-            # force the stop label of the baseline variant to 0
-            df.loc[(mname, '-', slice(None), 'stop'), ('value', self.baseline_variant)] = 0
-
-            if result.df is None:
-                result.df = df
-            else:
-                result.df = result.df.append(df)
+        worker = workerTable[method](**workerArgs)
+
+        result = {}
+        for kpi in self.reportKpiNames:
+            result[kpi] = {}
+            control       = self.getKPIbyNameAndVariant(kpi, self.controlVariantName)
+            controlWeight = self._getWeights(kpi, self.controlVariantName)
+            for variant in self.variantNames:
+                treatment       = self.getKPIbyNameAndVariant(kpi, variant)
+                treatmentWeight = self._getWeights(kpi, variant)
+                ds = worker(x=treatment*treatmentWeight, y=control*controlWeight)
+                result[kpi][variant] = {'controlVariant'   : self.controlVariantName,
+                                        'treatmentVariant' : variant,
+                                        'deltaStatistics'  : ds}
```

Comment:
Looking at the code, `worker` is a closure, containing either:
 - `statistics.delta()`, or
 - `early_stopping.group_sequential()`, or
 - `early_stopping.bayes_factor()`, or
 - `early_stopping.bayes_precision()`.

Looking at the definitions of those, you'll see that they basically return just a dictionary.